// @textbus/components/bundles/components/video.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var VideoComponent_1;
import { LeafAbstractComponent, VElement, Component } from '@textbus/core';
class VideoComponentLoader {
    constructor() {
        this.tagName = 'video';
    }
    match(component) {
        return component.nodeName.toLowerCase() === this.tagName;
    }
    read(el) {
        return {
            component: new VideoComponent(el.src, el.autoplay, el.controls, {
                width: el.style.width || el.width + '',
                height: el.style.height || el.height + ''
            }),
            slotsMap: []
        };
    }
}
let VideoComponent = VideoComponent_1 = class VideoComponent extends LeafAbstractComponent {
    constructor(src, autoplay, controls, option = {
        width: '100%',
        height: 'auto'
    }) {
        super('video');
        this.src = src;
        this.autoplay = autoplay;
        this.controls = controls;
        this.block = false;
        this.width = null;
        this.height = null;
        this.width = option.width;
        this.height = option.height;
    }
    render() {
        const el = new VElement(this.tagName);
        el.attrs.set('src', this.src);
        el.attrs.set('autoplay', this.autoplay);
        el.attrs.set('controls', this.controls);
        if (this.width) {
            el.styles.set('width', this.width);
        }
        if (this.height) {
            el.styles.set('height', this.height);
        }
        return el;
    }
    clone() {
        return new VideoComponent_1(this.src, this.autoplay, this.controls);
    }
};
VideoComponent = VideoComponent_1 = __decorate([
    Component({
        loader: new VideoComponentLoader()
    }),
    __metadata("design:paramtypes", [String, Boolean, Boolean, Object])
], VideoComponent);
export { VideoComponent };
