// @textbus/toolbar/bundles/toolbar/tools/table.tool.js
import { PreComponent } from '@textbus/components';
import { MenuType, GroupTool } from '../toolkit/_api';
import { tableAddToolConfig } from './table-add.tool';
import { tableEditToolConfig } from './table-edit.tool';
import { tableRemoveToolConfig } from './table-remove.tool';
import { tdBorderColorToolConfig } from './td-border-color.tool';
import { TableMatcher } from '../matcher/table.matcher';
export const tableToolConfig = {
    iconClasses: ['textbus-icon-table'],
    tooltip: i18n => i18n.get('plugins.toolbar.tableTool.tooltip'),
    matcher: new TableMatcher([PreComponent]),
    menu: [Object.assign(Object.assign({ type: MenuType.Dropdown }, tableAddToolConfig), { label: i18n => i18n.get('plugins.toolbar.tableTool.createTable') }), Object.assign(Object.assign({ type: MenuType.ActionSheet }, tableEditToolConfig), { label: i18n => i18n.get('plugins.toolbar.tableTool.editTable') }), Object.assign(Object.assign({ type: MenuType.Dropdown }, tdBorderColorToolConfig), { label: i18n => i18n.get('plugins.toolbar.tableTool.cellBorderColor') }), Object.assign(Object.assign({ type: MenuType.Action }, tableRemoveToolConfig), { label: i18n => i18n.get('plugins.toolbar.tableTool.deleteTable') })]
};
export const tableTool = new GroupTool(tableToolConfig);
