// @textbus/uikit/bundles/uikit/forms/form-textarea.js
export class FormTextarea {
    constructor(config) {
        this.config = config;
        this.elementRef = document.createElement('div');
        this.name = config.name;
        this.elementRef.classList.add('textbus-form-group');
        this.elementRef.innerHTML = `
    <div class="textbus-control-label">${config.label}</div>
    <div class="textbus-control-value">
      <div class="textbus-input-group textbus-input-block">
        <textarea name="${config.name}" class="textbus-form-control textbus-input-block" placeholder="${config.placeholder || ''}">${config.value || ''}</textarea>
     </div>
     <div class="textbus-control-feedback-invalid"></div>
   </div>`;
        this.input = this.elementRef.querySelector('textarea');
        this.feedbackEle = this.elementRef.querySelector('.textbus-control-feedback-invalid');
    }
    reset() {
        this.input.value = this.config.value;
    }
    update(value) {
        this.uploaded();
        this.input.value = (value !== null && value !== void 0 ? value : this.config.value) || '';
    }
    getAttr() {
        return {
            name: this.config.name,
            value: this.input.value
        };
    }
    validate() {
        var _a, _b;
        const feedback = (_b = (_a = this.config).validateFn) === null || _b === void 0 ? void 0 : _b.call(_a, this.getAttr().value);
        this.feedbackEle.innerText = feedback || '';
        return !feedback;
    }
    uploaded() {
        if (this.sub) {
            this.sub.unsubscribe();
        }
        this.input.disabled = false;
        if (this.btn) {
            this.btn.classList.remove('textbus-btn-loading');
            this.btn.children[0].className = 'textbus-icon-upload';
        }
    }
}
