// @textbus/uikit/bundles/uikit/forms/form-switch.js
export class FormSwitch {
    constructor(config) {
        this.config = config;
        this.elementRef = document.createElement('div');
        this.name = config.name;
        this.elementRef.classList.add('textbus-form-group');
        this.elementRef.innerHTML = `
    <div class="textbus-control-label"></div>
    <div class="textbus-control-static">
      <label><input name="${config.name}" type="checkbox" ${config.checked ? 'checked="checked"' : ''}> ${config.label}</label>
      <div class="textbus-control-feedback-invalid"></div>
    </div>
    `;
        this.input = this.elementRef.querySelector('input');
        this.feedbackEle = this.elementRef.querySelector('.textbus-control-feedback-invalid');
    }
    reset() {
        this.input.checked = this.config.checked;
    }
    update(value) {
        this.input.checked = typeof value === 'boolean' ? value : this.config.checked;
    }
    getAttr() {
        return {
            name: this.name,
            value: this.input.checked
        };
    }
    validate() {
        var _a, _b;
        const feedback = (_b = (_a = this.config).validateFn) === null || _b === void 0 ? void 0 : _b.call(_a, this.getAttr().value);
        this.feedbackEle.innerText = feedback || '';
        return !feedback;
    }
}
